### PR TITLE
assume_init: warn about valid != safe

### PR DESCRIPTION
**Pull request by RalfJung**
We have this warning in the type-level docs, but it seems worth repeating it on the function.

### CODE REVIEW

File: src/libcore/mem/maybe_uninit.rs [hunk 1]
```diff
@@ -402,6 +402,13 @@ impl<T> MaybeUninit<T> {
     ///
     /// [inv]: #initialization-invariant
     ///
+    /// On top of that, remember that most types have additional invariants beyond merely
+    /// being considered initialized at the type level. For example, a `1`-initialized [`Vec<T>`]
+    /// is considered initialized because the only requirement the compiler knows about it
+    /// is that the data pointer must be non-null. Creating such a `Vec<T>` does not cause
```
Comment (Mark-Simulacrum) [inline src/libcore/mem/maybe_uninit.rs]: Is this documented anywhere else? I'm not sure we want to settle on `Vec` being non-null as the only requirement.

**Comment by Mark-Simulacrum** [comment]
r=me either way

File: src/libcore/mem/maybe_uninit.rs (hunk 1 above)
Comment (RalfJung) [reply src/libcore/mem/maybe_uninit.rs]: This is a copy of text we already have in the `MaybeUninit` type-level docs.

I agree with your sentiment, not sure what would make a better example though.

Comment (Mark-Simulacrum) [reply src/libcore/mem/maybe_uninit.rs]: Yeah, I can't think of anything simple either. Maybe we could say creating `&u8` from `1` instead? That seems somehow simpler and doesn't get into `Vec` which is a pretty complex type from the user's point of view (generics, large API, etc.) even if all that doesn't matter for this case.

**Review by Centril — APPROVED** [review]

File: src/libcore/mem/maybe_uninit.rs (hunk 1 above)
Comment (Centril) [reply src/libcore/mem/maybe_uninit.rs]: Maybe use an example that is more domain specific? e.g. something involving a "proof of work" token that if misused triggers UB? //handwave-handwave

Comment (RalfJung) [reply src/libcore/mem/maybe_uninit.rs]: I think that would have to involve showing some code etc.

I was more thinking if we should weasle a bit more? Like, "with the current implementation, the all-1 bit pattern is legal (but this does not constitute a stable guarantee)".

Comment (Centril) [reply src/libcore/mem/maybe_uninit.rs]: Yeah I suppose that could work.

If you want something short-ish there's:

```rust
pub struct Id<S, T>(core::marker::PhantomData<(fn(S) -> S, fn(T) -> T)>);

pub fn refl<T>() -> Id<T, T> { Id(core::marker::PhantomData) }

impl<S, T> Id<S, T> {
    /// Casts a value of type `S` to `T`.
    pub fn cast(self, value: S) -> T {
        unsafe {
            // This is safe since we know by construction that
            // S == T (including lifetime invariance) always holds.
            let cast_value = core::mem::transmute_copy(&value);
            core::mem::forget(value);
            cast_value
        }
    }
}
```

which trivially satisfies the validity invariant but is crazy dangerous if you break the safety invariant.

Comment (RalfJung) [reply src/libcore/mem/maybe_uninit.rs]: Yeah I'd prefer to use a type the user is most likely already familiar with.

**Review by Centril — APPROVED** [review]
Looks good with the latest set of weasel words ;)

File: src/libcore/mem/maybe_uninit.rs [hunk 2]
```diff
@@ -402,6 +403,14 @@ impl<T> MaybeUninit<T> {
     ///
     /// [inv]: #initialization-invariant
     ///
+    /// On top of that, remember that most types have additional invariants beyond merely
+    /// being considered initialized at the type level. For example, a `1`-initialized [`Vec<T>`]
+    /// is considered initialized (under the current implementation, this does not constitute
```
Comment (bluetech) [inline src/libcore/mem/maybe_uninit.rs]: Grammatical note: there are two possible interpretations with the comma here. If you change the comma to a semicolon `;` or double dash `--`, is it the meaning you intended? If so, it will be better to change it. If not, it will be better to rephrase.

Another option is to drop the "this does not constitute a stable guarantee" part of the sentence because it is redundant with "under the current implementation".

Comment (Centril) [reply src/libcore/mem/maybe_uninit.rs]: > Another option is to drop the "this does not constitute a stable guarantee" part of the sentence because it is redundant with "under the current implementation".

It's not; it is making the point that the current implementation is not a stable guarantee, but I'd rephrase as: "under the current implementation, which does not constitute a stable guarantee".

Comment (RalfJung) [reply src/libcore/mem/maybe_uninit.rs]: @bluetech thanks! Turned it into a `;`
